### ISSUE TITLE
On startup, occasionally get "Failed to get 1 consecutive port..." message for slave instance

### ISSUE THREAD

**Issue by kdanderso** [sipwise/rtpengine#1503]
This is similar to issue #1495 

If a system is running rtpengine/redis with heavy call traffic and a new system starts in slave mode, the new system may be unable to open one or more requested ports for new calls. It appears that in this case, the affected port was opened by the redis_restore() function, but has not been released.

The failure occurs when redis is started as a slave and synchronized to the master before the slave instance of rtpengine is started.

I've attached a log from the slave system similar to that provided for #1495. These tests were run using commit 17bda4b

[rtpslave1.zip](https://github.com/sipwise/rtpengine/files/8903113/rtpslave1.zip)

Thank you.

**Comment by rfuchs** [comment]
Somehow the `del` notification for `3e803271-9d98-46a9-a11f-ac084e724ec4` is completely absent on the slave 🤔 Perhaps a race condition between reading in the existing calls and setting up the notifications... checking

**Comment by rfuchs** [comment]
So this is a bit of a shot in the dark, as I don't have a good way of reproducing or confirming the issue, but at least theoretically this should work, and superficial testing also looks good. My main concern is that this approach might lead to other unforeseen and unintended side effects, perhaps other race conditions. See dbd977c

**Comment by kdanderso** [comment]
So far, this seems to be working. I'll do more testing next week.

**Comment by kdanderso** [comment]
I am still seeing failures to get ports on startup, but the failure is not catastrophic. It looks to me like some of the messaging gets out of order so that a port is not released before it is requested again for a particular call. The port does not get permanently stuck, though, so other than causing a failure on a single call, there don't seem to be any long term effects. I've attached a log that shows this failure.

I am concerned that when the __get_consecutive_ports() function fails to get the requested start port, it tries to get another port from the free list, but then can't open the port because it's in use. In this case, that behavior has been advantageous, in that it revealed that the original port was not available. I have seen cases, though, where the port retrieved from the free list is opened, and the function returned successfully. Unfortunately, this means that a port has been reserved on the slave that the master doesn't know about. If the master subsequently tries to use that port, the slave will not be able to open it and may cause a cascading failure situation.

This prompts two questions that may need to be addressed: 1) Why is a port on the free list not actually available, and 2) should __get_consecutive_ports() even look for a port from the free list if wanted_start_port is non-zero.

[rtpslave2.zip](https://github.com/sipwise/rtpengine/files/9047956/rtpslave2.zip)

**Comment by rfuchs** [comment]
> Why is a port on the free list not actually available

Ports that are opened not randomly but by request (i.e. as it happens on the slave) bypass the free list. So the requested port is opened and is then in use, but remains on the free list. This is normal and expected and is why there is a retry mechanism in place.

> should __get_consecutive_ports() even look for a port from the free list if wanted_start_port is non-zero

No it should not, but one step at a time.

That failure from the log might be the opposite race condition now, a call being setup via notification first and then, while this is still ongoing, the restore thread comes across the same call at the same time and requests the port before it's released. Although I can't be 100% sure that this is really what happens just based on the log. I'll have to dig into it a bit more to see if this is actually possible.

**Comment by rfuchs** [comment]
Couple more commits here, somewhat experimental and yet completely untested as I ran out of time today, but I thought maybe you want to have an early look.

**Comment by kdanderso** [comment]
I tried these, but it looks like there's a deadlock on 0e8a761. In some cases, rtpengine just stops responding. Also, when it fails, SIGTERM does not terminate the rtpengine process. I've attached a log file in case it's helpful.

Just so you know, the failures I mentioned on dbd977c are fairly difficult to reproduce - I haven't been able to replicate them since the message I sent yesterday.

[rtpslave3.log](https://github.com/sipwise/rtpengine/files/9056306/rtpslave3.log)

**Comment by rfuchs** [comment]
Sorry, looks like there was a simply copy/paste-o in that commit. Now updated. New head is now  ca77494

**Comment by kdanderso** [comment]
It works occasionally, but tends to lock up after receiving a "Failed to restore call id" message. The message occurred in previous versions, but didn't cause lockups. Log attached.

[rtpslave4.zip](https://github.com/sipwise/rtpengine/files/9057794/rtpslave4.zip)

**Comment by rfuchs** [comment]
Ah deadlocks, always fun to deal with... Let's try that again with 5b72a3a6

**Comment by kdanderso** [comment]
This version looks good as far as I can tell from running tests overnight.

The testing I did uses two systems and a call generator. The first system uses redis keyspace one and is processing about 120,000 calls per hour. The second subscribes to keyspace one. I start redis and rtpengine on it, give them time to synchronize to the first system, and then see if there are any "Failed to get..." errors. If not, I shut down rtpengine and redis and repeat.

With dbd977c this test would typically fail within 10-20 passes, but sometimes could run longer (the longest I saw was 62 passes). The latest version has run over 4300 passes with no errors (and no deadlocks).

Thanks for your help.
